raspberry-pi/3/default.nix

```diff
@@ -0,0 +1,38 @@
+{ lib
+, pkgs
+, ...
+}:
+
+{
+  boot.kernelPackages = lib.mkDefault pkgs.linuxKernel.packages.linux_rpi3;
+
+  # fix the following error :
+  # modprobe: FATAL: Module ahci not found in directory
+  # https://github.com/NixOS/nixpkgs/issues/154163#issuecomment-1350599022
+  nixpkgs.overlays = [
+    (_final: super: {
+      makeModulesClosure = x:
+        super.makeModulesClosure (x // { allowMissing = true; });
+    })
+  ];
+
+  # https://github.com/NixOS/nixpkgs/blob/b72bde7c4a1f9c9bf1a161f0c267186ce3c6483c/nixos/modules/installer/sd-card/sd-image-aarch64.nix#L12
+  # Use the extlinux boot loader. (NixOS wants to enable GRUB by default)
+  boot.loader.grub.enable = lib.mkDefault false;
+  # Enables the generation of /boot/extlinux/extlinux.conf
+  boot.loader.generic-extlinux-compatible.enable = lib.mkDefault true;
+
+  # The serial ports listed here are:
+  # - ttyS0: for Tegra (Jetson TX1)
+  # - ttyAMA0: for QEMU's -machine virt
+  # https://github.com/NixOS/nixpkgs/blob/b72bde7c4a1f9c9bf1a161f0c267186ce3c6483c/nixos/modules/installer/sd-card/sd-image-aarch64.nix#L19
+  boot.kernelParams = [
+    "console=ttyS0,115200n8"
+    "console=ttyAMA0,115200n8"
+    "console=tty0"
+  ];
```

Comment:
```suggestion
  # The serial ports listed here are:
  # - ttyS0: serial
  # - tty0: hdmi
  boot.kernelParams = [
    "console=ttyS0,115200n8"
    "console=tty0"
  ];
```

Reply:
@pinage404 does your pi also shows tty0 even if there is no hdmi connected?
I am trying to figure out what the best default value is.

Reply:
With no HDMI plug-in

```sh
ll /dev/tty0
crw--w---- 1 root tty 4, 0  6 déc.   2023 /dev/tty0
```

With the current config

```nix
  boot.kernelParams = [
    "console=ttyS0,115200n8"
    "console=ttyAMA0,115200n8"
    "console=tty0"
  ];
```

Reply:
I don't think we ever need /dev/ttyAMA0 on raspberry pi. This device is qemu specific.
However printing kernel logs on hdmi is probably what most people expect. The configuration that I propose allows access via serial, but prints kernel logs to hdmi.

Reply:
i don't understand `boot.kernelParams`, it is black magic for me